src/utils/common.ts

```diff
@@ -290,33 +293,57 @@ export async function estimateTotalGasRequiredByUnsignedTransaction(
 }
 
 /**
- * Create an unsigned transaction of a fillRelay contract call
- * @param spokePool The specific spokepool that will populate this tx
- * @param destinationTokenAddress A valid ERC20 token (system-wide default is UDSC)
- * @param simulatedRelayerAddress The relayer address that relays this transaction
- * @returns A populated (but unsigned) transaction that can be signed/sent or used for estimating gas costs
+ * Create an unsigned transaction to fill a relay. This function is used to simulate the gas cost of filling a relay.
+ * @param spokePool A valid SpokePool contract instance
+ * @param fillToSimulate The fill that this function will use to populate the unsigned transaction
+ * @returns An unsigned transaction that can be used to simulate the gas cost of filling a relay
  */
-export async function createUnsignedFillRelayTransaction(
+export async function createUnsignedFillRelayTransactionFromFill(
   spokePool: SpokePool,
-  destinationTokenAddress: string,
-  simulatedRelayerAddress: string
+  fillToSimulate: Fill
 ): Promise<PopulatedTransaction> {
-  // Populate and return an unsigned tx as per the given spoke pool
-  // NOTE: 0xBb23Cd0210F878Ea4CcA50e9dC307fb0Ed65Cf6B is a dummy address
-  return await spokePool.populateTransaction.fillRelay(
-    "0xBb23Cd0210F878Ea4CcA50e9dC307fb0Ed65Cf6B",
-    "0xBb23Cd0210F878Ea4CcA50e9dC307fb0Ed65Cf6B",
-    destinationTokenAddress,
-    "10",
-    "10",
-    "1",
-    "1",
-    "1",
-    "1",
-    "1",
-    [],
+  const {
+    message,
+    recipient: recipientAddress,
+    relayer: relayerAddress,
+    amount: amountToRelay,
+    depositor: depositorAddress,
+  } = fillToSimulate;
+  const provider = spokePool.provider;
+
+  const isMessageEmpty = message.length === 0 || message === "0x";
+  // Only continue sanity checks if the message is not empty
+  if (!isMessageEmpty) {
+    const isRecipientAnAddress = ethers.utils.isAddress(recipientAddress);
+    const isRelayerAnAddress = ethers.utils.isAddress(relayerAddress);
+    if (!isRecipientAnAddress || !isRelayerAnAddress) {
+      throw new Error("Could not simulate message fill. Recipient address or relayer address is not a valid address");
+    }
+    const [isRecipientAContract, relayerBalanceOfToken] = await Promise.all([
+      isContractAddress(recipientAddress, provider),
+      getTokenBalance(relayerAddress, fillToSimulate.destinationToken, provider),
+    ]);
+    if (!isRecipientAContract) {
+      throw new Error("Could not simulate message fill. Recipient address is not a contract address");
+    }
```

Comment:
Why do we error in this case? Technically across allows messages where the recipient is not a contract.

Is the idea that this error will bubble up in the API preventing users from sending what looks like an erroneous transaction? If so, should that error be at a higher level (in the API code) rather than in the SDK where this code may be used for other things (like in the relayer codebase)?

Reply:
That's a fair point given how the message is essentially a no-op if the recipient is either not a contract. It was placed here to conform to the tech spec's version of graceful errors. I agree that there's a better place for this specific check and will make some revisions to move it 🙂 

```
relayExecution.updatedRecipient.isContract() && relayExecution.updatedMessage.length > 0
```

Reply:
yes so in this case this shouldn't be an error I agree. If the recipient is not a contract, then we should continue with the default behavior and simulating the fill with an empty message. In this case the message is ignored.

Reply:
Done: I've kicked the error up to the Vercel API where it matters to check for this case: https://github.com/across-protocol/frontend-v2/pull/862/commits/80cb76f5159902db393a33ce2fc5155571e4ba56